esphome/components/ltr_als_ps/ltr_als_ps.cpp

```diff
@@ -0,0 +1,520 @@
+  size_t i = size - 1;
+  size_t idx = -1;
+  while (idx == -1 && i > 0) {
+    if (array[i] == val) {
+      idx = i;
+      break;
+    }
+    i--;
+  }
+  if (idx == -1 || i == 0)
+    return val;
+  return array[i - 1];
+}
+
+static uint16_t get_itime_ms(IntegrationTime time) {
+  static const uint16_t ALS_INT_TIME[8] = {100, 50, 200, 400, 150, 250, 300, 350};
+  return ALS_INT_TIME[time & 0b111];
+}
+
+static uint16_t get_meas_time_ms(MeasurementRepeatRate rate) {
+  static const uint16_t ALS_MEAS_RATE[8] = {50, 100, 200, 500, 1000, 2000, 2000, 2000};
+  return ALS_MEAS_RATE[rate & 0b111];
+}
+
+static float get_gain_coeff(AlsGain gain) {
+  static const float ALS_GAIN[8] = {1, 2, 4, 8, 0, 0, 48, 96};
+  return ALS_GAIN[gain & 0b111];
+}
+
+static float get_ps_gain_coeff(PsGain gain) {
+  static const float PS_GAIN[4] = {16, 0, 32, 64};
+  return PS_GAIN[gain & 0b11];
+}
+
+void LTRAlsPsComponent::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up LTR-303/329");
+  // As per datasheet we need to wait at least 100ms after power on to get ALS chip responsive
+  this->set_timeout(100, [this]() { this->state_ = State::DELAYED_SETUP; });
+}
+
+void LTRAlsPsComponent::dump_config() {
+  auto get_device_type = [](LtrType typ) {
+    switch (typ) {
+      case LtrType::LTR_TYPE_ALS_ONLY:
+        return "ALS only";
+      case LtrType::LTR_TYPE_PS_ONLY:
+        return "PS only";
+      case LtrType::LTR_TYPE_ALS_AND_PS:
+        return "Als + PS";
+      default:
+        return "Unknown";
+    }
+  };
+
+  LOG_I2C_DEVICE(this);
+  ESP_LOGCONFIG(TAG, "  Device type: %s", get_device_type(this->ltr_type_));
+  if (this->is_als_()) {
+    ESP_LOGCONFIG(TAG, "  Automatic mode: %s", ONOFF(this->automatic_mode_enabled_));
+    ESP_LOGCONFIG(TAG, "  Gain: %.0fx", get_gain_coeff(this->gain_));
+    ESP_LOGCONFIG(TAG, "  Integration time: %d ms", get_itime_ms(this->integration_time_));
+    ESP_LOGCONFIG(TAG, "  Measurement repeat rate: %d ms", get_meas_time_ms(this->repeat_rate_));
+    ESP_LOGCONFIG(TAG, "  Glass attenuation factor: %f", this->glass_attenuation_factor_);
+    LOG_SENSOR("  ", "ALS calculated lux", this->ambient_light_sensor_);
+    LOG_SENSOR("  ", "CH1 Infrared counts", this->infrared_counts_sensor_);
+    LOG_SENSOR("  ", "CH0 Visible+IR counts", this->full_spectrum_counts_sensor_);
+    LOG_SENSOR("  ", "Actual gain", this->actual_gain_sensor_);
+  }
+  if (this->is_ps_()) {
+    ESP_LOGCONFIG(TAG, "  Proximity gain: %.0fx", get_ps_gain_coeff(this->ps_gain_));
+    ESP_LOGCONFIG(TAG, "  Proximity cooldown time: %d s", this->ps_cooldown_time_s_);
+    ESP_LOGCONFIG(TAG, "  Proximity high threshold: %d", this->ps_threshold_high_);
+    ESP_LOGCONFIG(TAG, "  Proximity low threshold: %d", this->ps_threshold_low_);
+    LOG_SENSOR("  ", "Proximity counts", this->proximity_counts_sensor_);
+  }
+  LOG_UPDATE_INTERVAL(this);
+
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "Communication with I2C LTR-303/329 failed!");
```

Comment:
If we change what's logged in `setup()`...just a callout to keep this consistent.
```suggestion
    ESP_LOGE(TAG, "Communication with I2C LTR-303/329/55x/659 failed!");
```